src/utils/MSFraggerAdapter.cpp

```diff
@@ -0,0 +1,859 @@
+    // and the working directory has been created
+    if (this->debug_level_ < 2 && this->working_directory != "")
+    {
+      File::removeDir(this->working_directory);
+    }
+  }
+
+
+protected:
+  void registerOptionsAndFlags_() final override
+  {
+    const StringList emptyStrings;
+    const std::vector< double > emptyDoubles;
+
+    const StringList validUnits = ListUtils::create<String>("Da,ppm");
+    const StringList isotope_error_and_enzyme_termini = ListUtils::create<String>("0,1,2");
+    const StringList zero_to_five = ListUtils::create<String>("0,1,2,3,4,5");
+
+    // Java executable
+    registerInputFile_(TOPPMSFraggerAdapter::param_java_executable, "<file>", "java", "The Java executable. Usually Java is on the system PATH. If Java is not found, use this parameter to specify the full path to Java", false, false, ListUtils::create<String>("skipexists"));
+    registerIntOption_(TOPPMSFraggerAdapter::param_java_heapmemory, "<num>", 3500, "Maximum Java heap size (in MB)", false);
+
+    // Handle executable
+    registerInputFile_(TOPPMSFraggerAdapter::param_executable, "<path_to_executable>", "", "Path to the MSFragger executable to use; may be empty if the executable is globally available.", false, false, ListUtils::create<String>("skipexists"));
+
+    // input spectra
+    registerInputFileList_(TOPPMSFraggerAdapter::param_in, "spectra_file_1 [spectra_file_2, .., spectra_file_N]>", emptyStrings, "Spectra files to search with MSFragger", true, false);
+    setValidFormats_(TOPPMSFraggerAdapter::param_in, ListUtils::create<String>("mzML,mzXML"));
+
+    //  output files
+    registerOutputFileList_(TOPPMSFraggerAdapter::param_out, "output_file_1 [output_file_2, ..., output_file_N]", emptyStrings, "MSFragger output files", true, false);
+    setValidFormats_(TOPPMSFraggerAdapter::param_out, ListUtils::create<String>("pep.xml,pepXML,tsv"), true);
+
+    // Path to database to search
+    registerInputFile_(TOPPMSFraggerAdapter::param_database, "<path_to_fasta>", "", "FASTA database file", true, false);
+    setValidFormats_(TOPPMSFraggerAdapter::param_database, ListUtils::create<String>("FASTA,fasta,fa,fas"), false);
+
+    // TOPP tolerance
+    registerTOPPSubsection_("tolerance", "Search Tolerances");
+
+    // Precursor mass tolerance and unit
+    _registerNonNegativeDouble(TOPPMSFraggerAdapter::param_precursor_mass_tolerance, "<precursor_mass_tolerance>", 20.0, "Precursor mass tolerance", false, false);
+    registerStringOption_(TOPPMSFraggerAdapter::param_precursor_mass_unit, "<precursor_mass_unit>", "ppm", "Unit of precursor mass tolerance", false, false);
+    setValidStrings_(TOPPMSFraggerAdapter::param_precursor_mass_unit, validUnits);
+
+    // Precursor true tolerance
+    _registerNonNegativeDouble(TOPPMSFraggerAdapter::param_precursor_true_tolerance, "<precursor_true_tolerance>", 0.0, "Precursor true tolerance. It is strongly recommended to set this feature!", false, false);
```

Comment:
a better description for this parameter would be nice, i guess.. I am only aware of precursor mass tolerance

Reply:
I improved the description of the parameters